sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java

```diff
@@ -504,14 +531,68 @@ private static void dryRunQuery(BigQueryOptions options, String query) {
 
       @Override
       public PCollection<TableRow> apply(PInput input) {
-        return PCollection.<TableRow>createPrimitiveOutputInternal(
-            input.getPipeline(),
-            WindowingStrategy.globalDefault(),
-            IsBounded.BOUNDED)
-            // Force the output's Coder to be what the read is using, and
-            // unchangeable later, to ensure that we read the input in the
-            // format specified by the Read transform.
-            .setCoder(TableRowJsonCoder.of());
+        String uuid = UUID.randomUUID().toString();
+        String jobIdToken = "beam_job_" + uuid;
+        String queryTempDatasetId = "temp_dataset_" + uuid;
+        String queryTempTableId = "temp_table_" + uuid;
+
+        BigQueryOptions bqOptions = input.getPipeline().getOptions().as(BigQueryOptions.class);
+
+        BoundedSource<TableRow> source;
+        BigQueryServices bqServices = getBigQueryServices();
+
+        final String extractDestinationDir;
+        String tempLocation = bqOptions.getTempLocation();
+        try {
+          IOChannelFactory factory = IOChannelUtils.getFactory(tempLocation);
+          extractDestinationDir = factory.resolve(tempLocation, uuid);
+        } catch (IOException e) {
+          throw new RuntimeException(
+              String.format("Failed to resolve extract destination directory in %s", tempLocation));
```

Comment:
Found I cannot use GcsPath directly.

In tests, the tempLocation can be either local or gcs, so I still need to use a IOChannelFactory.